plugin_tooling/src/com/github/rustdt/tooling/RustBuildOutputParserJson.java

```diff
@@ -0,0 +1,254 @@
+ * Copyright (c) 2016 Bruno Medeiros and other Contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Pieter Penninckx - initial implementation
+ *******************************************************************************/
+package com.github.rustdt.tooling;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+
+import melnorme.utilbox.status.Severity;
+
+import com.eclipsesource.json.Json;
+import com.eclipsesource.json.JsonArray;
+import com.eclipsesource.json.JsonObject;
+import com.eclipsesource.json.JsonValue;
+import com.eclipsesource.json.ParseException;
+
+import melnorme.lang.tooling.toolchain.ops.BuildOutputParser2;
+import melnorme.lang.tooling.toolchain.ops.OperationSoftFailure;
+import melnorme.lang.utils.parse.LexingUtils;
+import melnorme.lang.utils.parse.StringCharSource;
+import melnorme.utilbox.core.CommonException;
+import melnorme.utilbox.misc.IByteSequence;
+import melnorme.utilbox.process.ExternalProcessHelper.ExternalProcessResult;
+
+public abstract class RustBuildOutputParserJson extends BuildOutputParser2 {
+	
+	@Override
+	protected void handleNonZeroExitCode(ExternalProcessResult result) throws CommonException, OperationSoftFailure {
+		// Do nothing
+	}
+	
+	@Override
+	protected IByteSequence getOutputFromProcessResult(ExternalProcessResult result) {
+		return result.getStdErrBytes(); // Use StdErr instead of StdOut.
+	}
+	
+	// One line of Json data can contain more than one ToolMessageData, but we can only return one `ToolMessageData`
+	// from the method `parseMessageData`.
+	// So we create a buffer of remaining data.
+	protected String lineParsing = null; // The line we are currently working with. 
+	protected Queue<ToolMessageData> buffer; // Invariant: if buffer.isEmpty() then lineParsing == null
+	
+	public RustBuildOutputParserJson() {
+		super();
+		this.buffer = new ArrayDeque<ToolMessageData>();
+	}
+	
+	protected static final AbstractRustBuildOutputLineParser CARGO_OUTPUT_LINE_PARSER = 
+			new CargoRustBuildOutputLineParser();
+	
+	@Override
+	protected ToolMessageData parseMessageData(StringCharSource output) throws CommonException {
+		// The call-site stops calling `parseMessageData` when the `output` parameter is exhausted.
+		// So we may not completely exhaust the `output` variable when there are
+		// still `ToolMessageDatas`-s in the buffer.
+		ToolMessageData result = null;
+		try {
+			if (this.lineParsing != null) {
+				// Because of the invariant associated to buffer, we know that this.buffer is not empty.
+				result = this.buffer.remove();
+			} else {
+				this.lineParsing = LexingUtils.stringUntilNewline(output);
+				
+				if (this.lineParsing.startsWith("{")) {
```

Comment:
I don't quite get how this should work.
The catch for me is that while every output line of rustc is valid JSON, rustc's output as a whole is not. Simplified example to illustrate:

```json
{"message":"Cannot borrow immutable variable as mutable"}
{"message":"Cannot borrow variable as mutable more than once"}
```

This is not valid JSON, but rather two (since there are two lines) pieces of valid JSON joined together and separated by a newline. The newline character is used to separate two messages. 

Unfortunately, minimal-json cannot handle multiple JSON chunks separated by newlines. Let me illustrate this with an example:

```java
import com.eclipsesource.json.Json;
import com.eclipsesource.json.JsonValue;
import com.eclipsesource.json.ParseException;
import java.io.StringReader;
import java.io.IOException;

public class Main {
	public static void main(String[] args) {
		StringReader reader = new StringReader("{\"a\":1}\n{\"a\":2}");
		try {
			JsonValue a = Json.parse(reader);
		} catch (IOException ioe) {
			ioe.printStackTrace();
		}
	}
}
```

In this example, I create a Reader subclass that feeds something that is not quite JSON (but in the same style as rustc's output) to minimal-json. Instead of "eating" as much of the input as is valid JSON, `Json.parse()` throws a `com.eclipsesource.json.ParseException: Unexpected character at 2:1`.

In summary:

* `rustc`'s output as a whole is no valid JSON, but pieces of valid JSON separated by newlines
* minimal-json cannot handle this, so
* we split `rustc`'s output into lines (which are right now valid JSON) that we can feed to `Json.parse()`.

If future's `rustc`'s output were to split messages over multiple lines, the current approach would stop working as you noticed. I see different options to solve this:

1. Feed `rustc`'s output as a whole to `Json.parse()`. I believe this is what you proposed (correct me if I'm wrong), but unfortunately `Json.parse()` would throw an exception as illustrated with the example above.
2. Look for an alternative JSON parsing library that only eats as much as is valid JSON, parses that, and allows us to handle the rest (by feeding that to the same parser separately). 
3. Create a pull request for minimal-json to allow eating only as much as is valid JSON.
4. Override minimal-json's `Json.parse()` function (or `com.eclipsesource.json.JsonParser`'s method `parse(Reader reader, int buffersize)`).
5. Somehow detect which newlines are separating JSON parts and which are just whitespace. This can either be done by counting curly brackets (ignoring curly brackets which occur between quotes), or by trying to parse larger and larger pieces of the input until `Json.parse()` throws no exceptions anymore. I'm certainly no fan of the latter.
6. Cross our fingers and hope that `rustc`s output will not split JSON over multiple lines.

I would go for option 6., 3. or 4. What is your preference?

The only other IDE/editor that I found which uses the Json output is VisualRust. If I understood it correctly, they use `Newtonsoft.Json` (C#) which can [only eat as much as is valid json](http://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonReader_SupportMultipleContent.htm).